src/main/cpp/subsystems/Drivetrain.cpp

```diff
@@ -146,12 +151,28 @@ void Drivetrain::ControllerPeriodic() {
 
     UpdateDt();
 
-    Eigen::Vector<double, 5> y{
-        frc::AngleModulus(GetAngle()).value(), GetLeftPosition().value(),
-        GetRightPosition().value(), GetAccelerationX().value(),
-        GetAccelerationY().value()};
-    m_observer.Update(frc::Rotation2d(frc::AngleModulus(GetAngle())),
-                      GetLeftPosition(), GetRightPosition());
+    m_velPosObserver.Predict(m_u, GetDt());
+
+    m_leftPos = GetLeftPosition();
+    m_rightPos = GetRightPosition();
+    m_time = frc::Timer::GetFPGATimestamp();
+
+    auto rawLeftVelocity = (m_leftPos - m_lastLeftPos) / (m_time - m_lastTime);
+    auto rawRightVelocity = (m_rightPos - m_lastRightPos) / (m_time - m_lastTime);
+
+    m_leftVelocity = m_velocityFilter.Calculate(rawLeftVelocity);
+    m_rightVelocity = m_velocityFilter.Calculate(rawRightVelocity);
+
+    Eigen::Vector<double, 2> velPosY{GetLeftVelocity().value(),
+                                     GetRightVelocity().value()};
+
+    Eigen::Vector<double, 5> y{GetAngle().value(), GetLeftPosition().value(),
+                               GetRightPosition().value(),
+                               GetAccelerationX().value(),
+                               GetAccelerationY().value()};
+    m_observer.Update(GetAngle(), GetLeftPosition(), GetRightPosition());
+
+    m_velPosObserver.Correct(m_controller.GetInputs(), velPosY);
```

Comment:
You're supposed to use either the LinearFilter or the KalmanFilter, not both. The KalmanFilter you've made takes position measurements and contains position and velocity states.

Reply:
I changed it to be more like the flywheel so that the kalman filter is just a two state velocity filter based on the drivetrain plant. It seemed to work wonders today. So it takes the velocity in as the state. It's exactly what we did with the two flywheels.